% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/filterVCF_popstat.R
\name{get_frequency}
\alias{get_frequency}
\title{Get allele frequency from a filterVCF, VIF or VCFr object}
\usage{
get_frequency(VCF, Invariant = NULL, fixed = NULL,
  hets_prop_allowed = NULL, use_ff = TRUE)
}
\arguments{
\item{VCF}{a filterVCF object, a VIF object or a vcfR object}

\item{Invariant}{a Grange with invariant position on ly if VCF is of class "vcfR"}

\item{fixed}{a Grange with fixed position on ly if VCF is of class "vcfR"}

\item{hets_prop_allowed}{= NULL Threshold to remove sites with a certain proportion of heterozygous, intented for haploid.}

\item{use_ff}{= TRUE Should the genotype be ectracted using the ff library, this reduce the RAM memmory but is a bit slower}
}
\value{
\item{vcf_freq }{A Granges object for SNPs with frequencies as metadata} %% 
\item{invariant }{A Granges object for invariant with frequencies=0 as metadata} %%
\item{fixed }{A Granges object for fixed with frequencies=nb of chromosome sample in the population as metadata} %% 
...
}
\description{
If a filterVCF or VIF object is provided, invariant and fixed sites are automatically added to the frequency. Otherwise the user need to provide a Grange for both invariant and fixed sites. The function can use the ff library to reduce the allocated RAM used by the genotype matrix and the function extract_gt_ff() is used otherwise the more rapid extract.gt() from vcfR package is used.
}
\details{
The number of sample is taken from the vcfR object directly. the function returns a S3 list containing 3 GRanges element, snp frequecncy, invariant and fixed sites.
}
\examples{


freq <- get_frequency(filterVCF_object)
sapply(freq, length)

}
\references{

}
\author{
~~Benjamin Laenen~~
}
\keyword{popstat}
