% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/filterVCF_filters.R
\name{create_filter_fix_het_contiguous_in_pop}
\alias{create_filter_fix_het_contiguous_in_pop}
\title{Create a fiter based on heterozygous sites in samples}
\usage{
create_filter_fix_het_contiguous_in_pop(gt, vcf = NULL, pop_pattern,
  windows_size = 10, max_nb_contigous_hets = 2, report_hit = FALSE,
  ...)
}
\arguments{
\item{gt}{genotype matrix created with extract.gt or extract_gt_ff}

\item{vcf}{If the genotype is provided as ffdf, the vcfR object that what used to generate the matrix must be provided to extract the position. If not the vcf is not needed.}

\item{pop_pattern}{A regular expression to grep the names of the sample in the colnames of the genotype matrix. It can be a single pattern as "pop1.+" or a list of sample names "sample1|sample2|sampleN"}

\item{windows_size}{The size of the windows to detect the contiguous fixed heterozygous.}

\item{max_nb_contigous_hets}{The maximum number of contiguous fixed heterozygous allowed in the a windows.}

\item{...}{list of options in the form a list object: opt. Only opt$verbose}
}
\value{
GRanges object with windows with contiguous fixed heterozygous sites as TRUE
}
\description{
This function removes fixed heterozygous sites (i.e. all individuals are
0/1 for the sites) in a suite samples if those sites are in clustering
together. '   If at least 3 fixed heterozygous sites in the samples are
present in 10bp windows '   then the window is removed. This is another
filter intented '   to remove region with high repeats contents because fixed
heterozygous sites will '   likely be in small cluster (observed in IGV) if
coming from '   wrongly map reads. In order to identify the true wrongly map
read from legitimate fixed heterozygous the filter needs to be based on a
population that is expected to have very low level of heterozygosity. Highly
inbred or bottlenecked population could be used as the level of fixed
heterozygous is expected to be low and it is unlikely that if they are
present there will be clustering together.
}
\details{
This was especially designed for Arabis alpina
with the Swedish population as reference pop as they should have
much less fixed heterozygous sites being selfers.

Note of caution. Balancing selection could produce a pattern where a sites
is fixed for 2 alleles and can be trapped by this filter.
}
\examples{

##Select some samples and create a regex pattern
##sample_name <- colnames(gt)[c(1,2,5,8,6,12,15)]
##sample_pattern <- paste(sample_name, collapse = "|")
##filter <- create_filter_fix_het_contiguous_in_pop(gt, vcf = vcfR, pop_pattern = sample_pattern, windows_size = 10, max_nb_contigous_hets = 2)
##Remove the windows from the vcf
##vcf_filtered <- intersect_Granges(vcfR, filters, invert = TRUE)

}
\references{

}
\seealso{
objects to See Also as \code{\link{help}},

Other filters: \code{\link{change_gt_by_DP_filter}},
  \code{\link{change_gt_by_GQ_filter}},
  \code{\link{create_filter_META}},
  \code{\link{create_filter_all_het}},
  \code{\link{create_filter_fix_het}},
  \code{\link{create_filter_fixed_sites}},
  \code{\link{create_filter_high_DP_standardized}},
  \code{\link{create_filter_missing}},
  \code{\link{create_filter_repeats_in_windows}},
  \code{\link{create_indel_filter}},
  \code{\link{create_invariant_filter}},
  \code{\link{missing_per_individual}}
}
\author{
~~Benjamin Laenen~~
}
\concept{filters}
\keyword{~GATK}
\keyword{~filters}
