% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/filterVCF_vif_functions.R
\name{getSFS_VIF}
\alias{getSFS_VIF}
\title{Extract SFS, folded SFS and divergence from a VIF object}
\usage{
getSFS_VIF(vif, downsample_size = 20, divergence = FALSE,
  use_repol = FALSE, remove_polym_from_div = FALSE, freq = NULL)
}
\arguments{
\item{vif}{a VIF object}

\item{divergence}{= FALSE Should divergence be reported}

\item{use_repol}{= FALSE Is the VIF repolarized and should divergence be taken from the SFS?}

\item{remove_polym_from_div}{= FALSE remove polymorphism from divergence}

\item{freq}{= NULL provide a frequency object constructed using get_frequency(). This is meant to speed the retrieving of the SFS as the main speed bottleneck is getting the allele frequencies.}

\item{downsample_size=20}{report also a downsample SFS using downsample_sfs()}
}
\value{
A list
\item{SFS }{unfolded SFS} %% 
\item{fSFS }{folded SFS} %% 
\item{down_SFS }{downsampled unfolded SFS} %% 
\item{down_fSFS }{downsampled folded SFS} %% 
\item{D }{number of divergent sites between ingroup and outgroup(s)} %% 
\item{lD }{total number of divergence} %% 
...
}
\description{
The VIF object contains information on SNP, invariant and fixed sites by
default. Divergence can be added in the form of a GRange object with single
position and corresponding ingroup and outgroup(s) base.
}
\details{
The way the divergence is calculated depends wether the VIF object was
repolarized or not. If an outgroup is available and the VIF has been
repolarized using repolarize_VIF(), then divergence can be directly taken from the
SFS. The user can choose to count divergent sites as either all DERIVED sites
(including segregating) or only fixed DERIVED sites (instead of fixed
ALTERNATE to the reference when the VIF is not polarized). If an outgroup is
present but the VIF is not repolarized, the divergence can be calculated by
counting the number of sites that differ between the ingroup and outgroup
(NOTE that for it requires 2 outgroups and that the base is the same between
OUT1 and OUT2, it is hardcoded for the moment). It is still possible to
remove the SNPs from the divergence using this methods. The second methods
has the advantage of considering the whole genome alignment entirely while
the repolarized can only asses sites that were sequenced and called (SNP,
invariant and fixed). However, using repolarization offers a way to estimate
the probability that a sites is ancestral or derived.

The user is responsible to check that the VIF has been reporlarized and no check is
performed. Note that creating the repolKey does not change the VIF directly
but just create the mean to do so, the VIF needs to be repolarized in a
second step.
}
\examples{

getSFS_VIF(VIF_object) 

#include the divergence from a bed file in the VIF
divergence(VIF_object) <- bed2Grange("divergence.bed")
getSFS_VIF(VIF_object, divergence = TRUE) 


#make the repolarized key
repolKey(VIF_object)<- make_repol_key(VIF_object, methods = "ML")
repol_VIF_object <- repolarize_VIF(VIF_object)
getSFS_VIF(repol_VIF_object, divergence = TRUE, use_repol = TRUE) 


}
\author{
~~Benjamin Laenen~~
}
\keyword{vif}
