% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/filterVCF_popstat.R
\name{Popstat}
\alias{Popstat}
\title{Computed global population genetic statistics and window/range based}
\usage{
Popstat(filterVCF = NULL, vcf_file = NULL, invariant = NULL,
  fixed = NULL, reference = NULL, downsample_size = NULL,
  windows_size = NULL, windows_step = NULL, outputdir = NULL,
  output_name = NULL, sites = NULL, save2file = FALSE, ...)
}
\arguments{
\item{filterVCF}{A filterVCF or VIF object}

\item{vcf_file}{= NULL Unused if filterVCF is provided. path to a VCF file that can be read by read.vcfR()}

\item{invariant}{= NULL Unused if filterVCF is provided. A GRange with invariant sites.}

\item{fixed}{= NULL Unused if filterVCF is provided. A GRange with fixed sites.}

\item{reference}{= NULL Unused if filterVCF is provided. A DNAStringSet with the reference genome}

\item{downsample_size}{= NULL number of chromosome to downsample in the SFS. Note that the full SFS is always reported.}

\item{windows_size}{= NULL Size of the windows to compute statistics. If NULL only global stats is reported. If a GRange is provided, statistics will be calculated for each range (e.g. gene)}

\item{windows_step}{= NULL size of the sliding windows. If NULL non-overlapping windows is used}

\item{outputdir}{= NULL Name of the output directory if save2file = TRUE}

\item{output_name}{= NULL Root name of the output  if save2file = TRUE}

\item{sites}{= NULL A GRange to intersect with. For example, the user can provide non-synonymous sites and computes stats per windows or per genes.}

\item{save2file}{= FALSE Save output to a bed file with extra columns?}

\item{...}{additional options from initialise_option()}
}
\value{
A list with 3 elements
\item{vcf_freq }{A vcf_freq object}
\item{global_stats }{A list with the global statistics}
\item{stats_windows }{A Grange with stats in the metadata}
}
\description{
This function is the main utility to compute population genetic statsitics
using the filterVCF of VIF object class from this package. However, users can
also read in their own VCF file but must provide invariant sites as the main
purpose of this package is to always consider all sites that have bee
actually sequence when computing the statistic to avoid inflating value dues
to lack of sequencing in some genomic regions.
}
\details{
Global statistics will be calculated for the whole object. User can also
compute statisticsby windows based on the reference genome present in the
filterVCF or VIF object using the function get_stats_windows(). 
If a GRange is provided instead of window size, then
the statistics is calculated for each range (e.g. gene, exon, 4fold).

#' Nucleotide diversity Pi
	Is calculated per sites using corrected Begun (2007) formula and divided by the number of sites actually sequenced. (Note : there was an error in the denominator in the original paper): 
		pi <- 2 * (freq ALT * (nsample - freq ALT)) / (nsample * (nsample -1))
	Begun, D. J., Holloway, A. K., Stevens, K., Hillier, L. W., Poh, Y. P., Hahn, M. W., ... & Pachter, L. (2007). Population genomics: whole-genome analysis of polymorphism and divergence in Drosophila simulans. PLoS biology, 5(11), e310.

Whaterson's theta
Is calculated as follow and divided by the number of sites actually sequenced:	
		S / sum(1 / 1:(nsample - 1))
	Watterson, G. A. (1975). On the number of segregating sites in genetical models without recombination. Theoretical population biology, 7(2), 256-276.

Tajima's D
	The numerator 'd' is calculated as d = sum(pi) - (S / sum(1 / 1:(nsample - 1))). 'd' is then passed too pegas::tajima.test() to get D. Note that for Tajima's D the total number of sites is irelevant as it does the difference between two estimator of theta this is unscaled.
	Tajima, F. (1989). Statistical method for testing the neutral mutation hypothesis by DNA polymorphism. Genetics, 123(3), 585-595.

Fu's Fs 
	Not yet implemented. (make a git merge request if you have some code)

##NOT TESTED YET## used it with care	
Fay and Wu H
	Calculated as follow:
	thetaH  <- sum(SFS[-c(1,length(SFS))] * (1:(nsample-1))^2) / choose(nsample,2)
	H <- sum(pi) - thetaH
Fay, J. C., & Wu, C. I. (2000). Hitchhiking under positive Darwinian selection. Genetics, 155(3), 1405-1413.

##NOT TESTED YET## used it with care	
	thetaL  <- sum(SFS[-c(1,length(SFS))] * (1:(nsample-1))) / (nsample-1)^-1
	E <- thetaL - thetaW
K. Zeng, Y.-X. Fu, S. Shi, and C.-I. Wu. Statistical tests for detecting positive selection by utilizing high-frequency variants. Genetics, 174:1431â€“1439, 2006.
}
\examples{

#Compute only global statistics
vif_object <- as.VIF(filterVCF_object)
Popstat(vif_object , windows_size = NULL, windows_step = NULL, outputdir = NULL, output_name = NULL, sites = NULL, save2file = FALSE)
 
#Compute only global statistics for synonymous
#Read in synonymous sites
 syn <- bed2Grange("4fold_syn.bed")
Popstat(vif_object , windows_size = NULL, windows_step = NULL, outputdir = NULL, output_name = NULL, sites = syn, save2file = FALSE)
 
#Compute statistics for synonymous per gene and save to a bed file
#Read in gff and selct only gene
gff <- import("my_species.gff3", format = "gff")
gff <- gff[values(gff)$type == "gene"]

Popstat(vif_object , windows_size = gff, windows_step = NULL, outputdir = "synonymous_stats", output_name = "my_species_syn_gene", sites = syn, save2file = TRUE)

#Compute statistics for all sites in 20kb windows and save to a bed file
Popstat(vif_object , windows_size = 20000, windows_step = NULL, outputdir = "all_sites_stats", output_name = "my_species_20kb", sites = NULL, save2file = TRUE)

# Use external VCF file from the user.
vcf_file = "my.vcf"
invariant <- bed2Grange("invariant_sites.bed")
#Note that it is not necesaary if they are still present in the VCF
fixed <- bed2Grange("fixed_sites.bed")
reference <- readDNAStringSet("my.genome.fasta")

Popstat(filterVCF = NULL ,  vcf_file = vcf_file, invariant = invariant, fixed = fixed, reference = reference, windows_size = 20000, windows_step = NULL, outputdir = "my_vcf_stats", output_name = "my_vcf_20kb", sites = NULL, save2file = TRUE)

}
\references{

}
\author{
~~Benjamin Laenen~~
}
\keyword{~popstat}
