% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/filterVCF_filters.R
\name{create_filter_all_het}
\alias{create_filter_all_het}
\title{Create a filter for all heterozygous sites}
\usage{
create_filter_all_het(gt, ...)
}
\arguments{
\item{gt}{genotype matrix created with extract.gt or extract_gt_ff}

\item{...}{list of options in the form a list object: opt. Only opt$verbose}
}
\value{
logical vector of sizes nrow(gt) with sites not passing the filter as TRUE
}
\description{
The function will create a logical filter the same length as the vcf with
all heterozygous sites or all heterozygous and missing. This must be used
with care as it can biased the resulting data. This is indetended for haploid
to catch wrongly mapped reads.
}
\note{
~~further notes~~
}
\examples{

##opt <- list(verbose=TRUE)
##filter <- create_filter_all_het(gt, opt)
##sum(filter)

}
\references{

}
\seealso{
objects to See Also as \code{\link{help}},

Other filters: \code{\link{change_gt_by_DP_filter}},
  \code{\link{change_gt_by_GQ_filter}},
  \code{\link{create_filter_META}},
  \code{\link{create_filter_fix_het_contiguous_in_pop}},
  \code{\link{create_filter_fix_het}},
  \code{\link{create_filter_fixed_sites}},
  \code{\link{create_filter_high_DP_standardized}},
  \code{\link{create_filter_missing}},
  \code{\link{create_filter_repeats_in_windows}},
  \code{\link{create_indel_filter}},
  \code{\link{create_invariant_filter}},
  \code{\link{missing_per_individual}}
}
\author{
~~Benjamin Laenen~~
}
\concept{filters}
\keyword{filter}
