% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/filterVCF_convert.R
\name{extract_gt_ff}
\alias{extract_gt_ff}
\title{Extract genotype using memmory efficient ffdf data.frame}
\usage{
extract_gt_ff(vcf, ...)
}
\arguments{
\item{vcf}{A vcfR object}

\item{...}{list of options in the form a list object: opt. Only opt$verbose
is used here. Default is true.}
}
\value{
a ffdf with genotype
}
\description{
The genotype matrix exported by the vcfR package can be quite large in RAM
memmory and this function aimed to circumvent this issue by using the ff
package to use memory efficiently. The genotypes are extracted from the VCF
field GT wheter they are phased or not but phase information is not retained.
The genotype are coded as 
-9 = missing
1 = heterozygous
0 = homozygous for REF allele
2 = homozygous for ALT allele
}
\details{
~~ If necessary, more details than the description above ~~
}
\note{
~~further notes~~
}
\examples{


gt <- extract_gt_ff(vcfR)
gt

}
\references{
http://ff.r-forge.r-project.org/
}
\seealso{
Other utilities VCFutilities: \code{\link{vcf2Grange}}
}
\author{
~~Benjamin Laenen~~
}
\concept{utilities VCFutilities}
\keyword{~utilities}
