% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/filterVCF_popstat.R
\name{get_theta_pi_SFS}
\alias{get_theta_pi_SFS}
\title{Compute population genetic statistics and site frequency spectrum (SFS)}
\usage{
get_theta_pi_SFS(vcf_freq, invariant, fixed, downsample_size = 20,
  output_stats_only = TRUE, n = NULL, SFS_only = FALSE)
}
\arguments{
\item{vcf_freq}{a vcf_frequency object obtain using get_frequency() or GRanges for SNPs with frequency as value.}

\item{invariant}{GRange for invariant, not used if a vcf_freq object from get_frequency() is provided}

\item{fixed}{GRange for fixed, not used if a vcf_freq object from get_frequency() is provided}

\item{downsample_size}{= 20 Size of the downsampling}

\item{SFS_only}{= FALSE report only the SFS}

\item{output_stats_only=TRUE}{return only the stats, if FALSE return a vcf_freq object and the stats}

\item{n=NULL}{Number of sample, if NULL it is taken from the vcf_freq. NOTE that this parameter influence all statistics, change it only if you know what you are doing!}
}
\value{
A list with statistics
\item{total_nb_sites }{Total number of sites sequenced, denominator of nuc_div and thetaW} 
\item{S }{Number of segragating sites} 
\item{REF_fixed }{Total number of invariant sites for the REF allele} 
\item{ALT_fixed }{Total number of fixed sites for the ALT allele} 
\item{SFS }{list containing  unfolded, folded and downsampled SFSs} 
\item{nuc_div }{Nucleotide diversity Pi relative to the total number of sites sequenced} 
\item{thetaW }{Whaterson's theta relative to the total number of sites sequenced} 
\item{tajimaD }{Tajima's D} 
\item{FayWu_H }{Fay and Wu H} 
\item{Zeng_E }{Zeng E}
}
\description{
This function is used internally in Popstat() to compute, nucleotide
diversity, Whaterson's theta, Tajima's D, Fay and Wu H, and Zeng E. The SFS
is also extracted considering all sites that were sequenced including SNPs,
invariant and fixed and return the unfolded, folded and downsampled SFS. Note
that for unfolded SFS the user must first repolarize the vif object using
make_repolKey() and repolarize_VIF(). This function is used to calculate stats
on subset of the genome, such as gene, loci, windows and will always consider
the actual number of sites examined.
}
\details{
Nucleotide diversity Pi
	Is calculated per sites using corrected Begun (2007) formula and divided by the number of sites actually sequenced. (Note : there was an error in the denominator in the original paper): 
		pi <- 2 * (freq ALT * (nsample - freq ALT)) / (nsample * (nsample -1))
	Begun, D. J., Holloway, A. K., Stevens, K., Hillier, L. W., Poh, Y. P., Hahn, M. W., ... & Pachter, L. (2007). Population genomics: whole-genome analysis of polymorphism and divergence in Drosophila simulans. PLoS biology, 5(11), e310.

Whaterson's theta
Is calculated as follow and divided by the number of sites actually sequenced:	
		S / sum(1 / 1:(nsample - 1))
	Watterson, G. A. (1975). On the number of segregating sites in genetical models without recombination. Theoretical population biology, 7(2), 256-276.

Tajima's D
	The numerator 'd' is calculated as d = sum(pi) - (S / sum(1 / 1:(nsample - 1))). 'd' is then passed too pegas::tajima.test() to get D. Note that for Tajima's D the total number of sites is irelevant as it does the difference between two estimator of theta this is unscaled.
	Tajima, F. (1989). Statistical method for testing the neutral mutation hypothesis by DNA polymorphism. Genetics, 123(3), 585-595.

Fu's Fs 
	Not yet implemented. (make a git merge request if you have some code)

##NOT TESTED YET## used it with care	
Fay and Wu H
	Calculated as follow:
	thetaH  <- sum(SFS[-c(1,length(SFS))] * (1:(nsample-1))^2) / choose(nsample,2)
	H <- sum(pi) - thetaH
Fay, J. C., & Wu, C. I. (2000). Hitchhiking under positive Darwinian selection. Genetics, 155(3), 1405-1413.

##NOT TESTED YET## used it with care	
	thetaL  <- sum(SFS[-c(1,length(SFS))] * (1:(nsample-1))) / (nsample-1)^-1
	E <- thetaL - thetaW
K. Zeng, Y.-X. Fu, S. Shi, and C.-I. Wu. Statistical tests for detecting positive selection by utilizing high-frequency variants. Genetics, 174:1431â€“1439, 2006.
}
\examples{

freq <- get_frequency(filterVCF_object)
get_theta_pi_SFS(freq, downsample_size = 10)

}
\references{

}
\author{
~~Benjamin Laenen~~
}
\keyword{~popstat}
