% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/filterVCF_popstat.R
\name{get_stats_windows}
\alias{get_stats_windows}
\title{Compute pi , Wahterson theta, Tajima's D, Fu's Fs, Fay and Hu H and SFS in windows or per genomic ranges}
\usage{
get_stats_windows(vcf_freq, reference, windows_size = 20000,
  step = 20000, downsample_size = 20)
}
\arguments{
\item{vcf_freq}{a vcf_freq object obtained qith get_frequency}

\item{reference}{A DNAStringSet with the reference genomes or the path to the fasta reference genome. If windows size is a GRanges set reference = NULL.}

\item{step}{= 20000 Size of the sliding in bp. Set step = windows_size for non-overlapping windows and ignore if windows_size is a GRange}

\item{downsample_size}{= 20 size of the downsampling of the SFS. Note that is only optional but the full SFS is always outputed.}

\item{windows_size=20000}{Size of the windows in bp. Alternatively A Granges with for wichi stats should be computed on each interval.}
}
\value{
A Granges object each range corresponding to either a window or a range from the inputed GRange (see above) with associated stats and information about the number of sites, total, segragating and fixed and the SFS
}
\description{
This function call get_theta_pi_SFS() in windows or on each element of a Grange. Thus it can output stats for separate genes or any ranges defined in windows size. See get_theta_pi_SFS() for more info on the statistic computed
}
\examples{

#Get the allele frequencies
freq <- get_frequency(filterVCF_object)

#import the gff and keep only CDS
gff_file <- "yourspecies.gff3"
gff <- import(gff_file, format = "gff")
gff <- gff[values(gff)[,"type"] == "CDS"] 


#Get the statistic for each CDS
CDS_stats <- get_stats_windows(freq, reference(filterVCF_object), windows_size = gff)

#Get the statistic in 50kb windows
windows_stats <- get_stats_windows(freq, reference(filterVCF_object), windows_size = 50000, step = 50000)

}
\references{

}
\author{
~~Benjamin Laenen~~
}
