% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/filter_VCF_utilities.R
\name{remove_overlapping_range}
\alias{remove_overlapping_range}
\title{Remove overlaping ranges}
\usage{
remove_overlapping_range(Grange, ...)
}
\arguments{
\item{Grange}{A GRanges object}

\item{...}{argument ignore.strand can be passed to findOverlaps to ignore strandness inoverlap}
}
\value{
A GRanges object with the smallest overlapping interval(s) removed
}
\description{
Function to test and remove ranges that are 
overlapping by keeping the longest ranges and 
keeping the metadata.
}
\details{
This function is intended to remove overlapping gene
and keep the longest (primary) transcript. If a GFF is
read into a GRanges, it can remove all the exon/CDS parts but
keeping the gene definition.
}
\examples{

Grange <- GRanges(seqnames="chr1", ranges= c(1:10, 1:4, 2:5, 1:2, 100:1000, 120:140)) 
Grange_simplified <- remove_overlapping_range(Grange)

Grange <- GRanges(seqnames="chr1", ranges= c(1:10, 1:4, 2:5, 1:2, 100:1000, 120:140), strand = c("+", "+", "-", "-", "+", "+")) 
Grange_simplified <- remove_overlapping_range(Grange)
# Ignoring strand
Grange_simplified <- remove_overlapping_range(Grange, ignore.strand = TRUE)

}
\references{
put references to the literature/web site here
}
\author{
~~Benjamin Laenen~~
 objects to See Also as \code{\link{help}},
}
\keyword{~utilities}
